server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,258 @@
+
+         synchronizer.resultFuture.whenComplete((ignore_v, ignore_t) -> {
+            deleteTimer();
+            cache.removeListenerAsync(this);
+         });
+      }
+
+      public CompletableFuture<Collection<byte[]>> getFuture() {
+         return synchronizer.resultFuture;
+      }
+
+      private void startTimer(long timeout) {
+         deleteTimer();
+         scheduledTimer = (timeout > 0) ? handler.getScheduler().schedule(() -> {
+            cache.removeListenerAsync(this);
+            synchronizer.resultFuture.complete(null);
+         }, timeout, TimeUnit.MILLISECONDS) : null;
+      }
+
+      private void deleteTimer() {
+         if (scheduledTimer != null)
+            scheduledTimer.cancel(true);
+         scheduledTimer = null;
+      }
+
+      @CacheEntryCreated
+      @CacheEntryModified
+      public void onEvent(CacheEntryEvent<Object, Object> entryEvent) {
+         try {
+            if (entryEvent.getValue() instanceof ListBucket) {
+               byte[] key = unwrapKey(entryEvent.getKey());
+               synchronizer.onEvent(key);
+            }
+         } catch (Exception ex) {
+            synchronizer.resultFuture.completeExceptionally(ex);
+         }
+      }
+
+      private byte[] unwrapKey(Object key) {
+         return key instanceof WrappedByteArray
+               ? ((WrappedByteArray) key).getBytes()
+               : (byte[]) key;
+      }
+   }
+
+   /**
+    * PollListenerSynchronizer
+    *
+    * This class synchronizes the access to a CompletableFuture `resultFuture` so
+    * that its final value will be completed either
+    * - with value v by an onListenerAdded() call if a not null value is found
+    * - otherwise by onEvent(k) call if the referred entry is not null or empty
+    *
+    */
+   public static class PollListenerSynchronizer {
+      private final ConcurrentLinkedQueue<Object> keyQueue;
+      private final CompletableFuture<Collection<byte[]>> resultFuture;
+      private final EmbeddedMultimapListCache<byte[], byte[]> multimapList;
+      private final List<byte[]> keys;
+      private final BiConsumer<? super Collection<byte[]>, ? super Throwable> whenCompleteConsumer;
+      private boolean canPollJustEventKey;
+
+      private PollListenerSynchronizer(List<byte[]> keys, EmbeddedMultimapListCache<byte[], byte[]> multimapList) {
+         keyQueue = new ConcurrentLinkedQueue<Object>();
+         resultFuture = new CompletableFuture<Collection<byte[]>>();
+         this.multimapList = multimapList;
+         this.keys = keys;
+         whenCompleteConsumer = (v, t) -> {
+            if (t != null) {
+               resultFuture.completeExceptionally(t);
+            } else if (v != null && !v.isEmpty()) {
+               resultFuture.complete(v);
+            } else {
+               if (keyQueue.poll() == this) {
+                  canPollJustEventKey = true;
+               }
+               if (!keyQueue.isEmpty()) {
+                  runPoll();
+               }
```

Comment:
```suggestion
               byte[] key;
               synchronized (this) {
                 if (keyQueue.poll() == this) {
                     canPollJustEventKey = true;
                  }
                  key = keyQueue.peek();
               }

               if (key != null) {
                  runPoll(key);
               }
```

This needs to be protected by the synchronized block as well. Just wrap the outer if in synchronized(this) block should be all you need.